rust/main/hyperlane-metric/src/sealevel/prometheus_sender.rs

```diff
@@ -0,0 +1,160 @@
+        }
+        if let Some(counter) = &self.metrics.request_duration_seconds {
+            counter
+                .with(&labels)
+                .inc_by((Instant::now() - start).as_secs_f64())
+        };
+        res
+    }
+    fn url(&self) -> String {
+        self.inner.url.clone()
+    }
+}
+
+async fn send_sealevel_rpc_request(
+    client: &Arc<reqwest::Client>,
+    url: String,
+    request_id: u64,
+    request: RpcRequest,
+    params: serde_json::Value,
+) -> Result<serde_json::Value, ClientError> {
+    let request_json = {
+        let jsonrpc = "2.0";
+        serde_json::json!({
+            "jsonrpc": jsonrpc,
+            "id": request_id,
+            "method": format!("{}", request),
+            "params": params,
+        })
+        .to_string()
+    };
+
+    let mut too_many_requests_retries = 5;
+
+    loop {
+        let response = {
+            let request_json = request_json.clone();
+            client
+                .post(&url)
+                .header(CONTENT_TYPE, "application/json")
+                .body(request_json)
+                .send()
+                .await
+        }?;
+
+        if !response.status().is_success() {
+            if response.status() == StatusCode::TOO_MANY_REQUESTS && too_many_requests_retries > 0 {
+                let mut duration = Duration::from_millis(500);
+                if let Some(retry_after) = response.headers().get(RETRY_AFTER) {
+                    if let Ok(retry_after) = retry_after.to_str() {
+                        if let Ok(retry_after) = retry_after.parse::<u64>() {
+                            if retry_after < 120 {
+                                duration = Duration::from_secs(retry_after);
+                            }
+                        }
+                    }
+                }
+
+                too_many_requests_retries -= 1;
+                sleep(duration).await;
+                continue;
+            }
+            return Err(response.error_for_status().unwrap_err().into());
+        }
+
+        let mut json = response.json::<serde_json::Value>().await?;
+        if json["error"].is_object() {
+            return match serde_json::from_value::<RpcErrorObject>(json["error"].clone()) {
+                Ok(rpc_error_object) => {
+                    let data = rpc_error_object_to_response(&json, &rpc_error_object);
+                    Err(RpcError::RpcResponseError {
+                        code: rpc_error_object.code,
+                        message: rpc_error_object.message,
+                        data,
+                    }
+                    .into())
+                }
+                Err(err) => Err(RpcError::RpcRequestError(format!(
+                    "Failed to deserialize RPC error response: {} [{}]",
+                    serde_json::to_string(&json["error"]).unwrap(),
+                    err
+                ))
+                .into()),
+            };
+        }
+        return Ok(json["result"].take());
```

Comment:
this looks pretty complicated, do we have to do this parsing and retrying manually here? My expectation is that we can wrap a provider that does all this for us, and we just collect usage metrics, like we do for the ethers provider

Reply:
Yeah initially I was planning to use HttpSender from solana-client, but its private. So I ended up copying most of the code over. I'll add documentation on where the code comes from